src/execution/mod.rs

```diff
@@ -342,6 +344,260 @@ where
+                                                    let mut stack = Stack::new();
+                                                    // TODO: fully implement run_const
+                                                    run_const(wasm, &mut stack, (), &function_instances);
+                                                    let value = stack.peek_unknown_value();
+                                                    if value.is_none() {
+                                                        panic!("No value on the stack for element segment offset");
+                                                    }
+                                                    value.unwrap()
+                                                };
+
+                                                let offset: u32 = match value {
+                                                    // Value::I32(val) => val,
+                                                    // Value::I64(val) => {
+                                                    //     if val > u32::MAX as u64 {
+                                                    //         panic!("i64 value for data segment offset is out of reach")
+                                                    //     }
+                                                    //     val as u32
+                                                    // }
+                                                    // INFO: no need to implement all of them, it's either i32 or i64, otherwise offset is WRONG
+                                                    // INFO2: wait, we might need globals handling, now that I think about it, but make it return an u32 anyways
+                                                    Value::Ref(rref) => {
+                                                        match rref {
+                                                            Ref::Func(_) => unreachable!(),
+                                                            Ref::Extern(extern_addr) => extern_addr.addr as u32
+                                                        }
+                                                    }
+                                                    _ => {
+                                                        unreachable!()
+                                                    },
+                                                };
+
+                                                Ref::Extern(ExternAddr::new(Some(offset as usize)))
+                                            }).collect::<Vec<Ref>>()
+                                        },
+                                        ElemItems::RefFuncs(_) => panic!("RefFuncs allowed only for Functions!")
+                                    }
+                                }
+                            },
+                        })
+                    },
+                    ElemMode::Active(active_elem) => {
```

Comment:
This feels quite similar to the implementation of ElemMode::Passive. Could we perhaps merge the two and the differences (if there aren't a lot) to be handled with a handful of if let / small nested match statements between the common code